torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,1360 @@
+  """
+
+  def __init__(
+      self,
+      module: nn.Module,
+      reshard_after_forward: bool = True,
+      flatten_parameters: bool = True,
+      execute_sharding_on_init: bool = True,
+      optimization_barrier_in_forward: bool = True,
+      optimization_barrier_in_backward: bool = True,
+      use_all_gather_via_all_reduce: bool = False,
+      mark_step_on_freeing: bool = False,
+      mark_step_on_finalization: bool = False,
+      _debug_dummy_forward_pass: bool = False,
+      _debug_msg: str = "xla_fsdp",
+      _debug_print: bool = False,
+      _shard_size_multiple: int = 128,
+      _pin_layout_in_all_reduce: bool = False,
+      _pin_layout_in_all_gather: bool = False,
+      _pin_layout_in_reduce_scatter: bool = False,
+      _debug_dummy_all_gather_op: bool = False,
+      _debug_dummy_reduce_scatter_op: bool = False,
+  ):
+    if isinstance(module, XlaFullyShardedDataParallel):
+      raise RuntimeError(
+          "Cannot wrap a module that is already wrapped with FSDP. For nested FSDP, "
+          "first wrap the inner child modules before wrapping the outer parent module."
+      )
+    is_forward_defined = (
+        hasattr(module, "forward") and hasattr(module.forward, "__func__") and
+        module.forward.__func__ != torch.nn.Module.forward)
+    if not is_forward_defined:
+      raise RuntimeError(
+          "The module wrapped by FSDP *must define a `forward` method and call it "
+          "during the module's forward pass for FSDP to work correctly.* "
+          "Hence, do not wrap `nn.ModuleList` or `nn.ModuleDict` with FSDP "
+          "(since they don't have `forward` defined), "
+          "and do not perform the forward pass in other ways apart from the `forward` method. "
+          "(i.e. you should directly call the FSDP-wrapped module itself in your code, "
+          "instead of using any of its submodules or its weights).")
+
+    super().__init__()
+    self.rank = xm.get_ordinal()
+    self.world_size = xm.xrt_world_size()
+    self.reshard_after_forward = self._orig_reshard_after_forward = reshard_after_forward
+    self.flatten_parameters = flatten_parameters
+    self.optimization_barrier_in_forward = optimization_barrier_in_forward
+    self.optimization_barrier_in_backward = optimization_barrier_in_backward
+    if _debug_dummy_all_gather_op:
+      self.all_gather_op = dummy_all_gather
+    elif use_all_gather_via_all_reduce:
+      self.all_gather_op = functools.partial(
+          all_gather_via_all_reduce, pin_layout=_pin_layout_in_all_reduce)
+    else:
+      self.all_gather_op = functools.partial(
+          xm.all_gather, pin_layout=_pin_layout_in_all_gather)
+    if _debug_dummy_reduce_scatter_op:
+      self.reduce_scatter_op = dummy_reduce_scatter
+    else:
+      self.reduce_scatter_op = functools.partial(
+          xm.reduce_scatter, pin_layout=_pin_layout_in_reduce_scatter)
+    # TODO (ronghanghu): remove when https://github.com/pytorch/xla/issues/3455 is resolved
```

Comment:
This issue is resolved at this point. Shall we handle this TODO prior to merge?

Reply:
(this obsolete option is now removed)